cpp/src/arrow/compute/kernels/aggregate_basic.cc

```diff
@@ -218,6 +218,65 @@ Result<std::unique_ptr<KernelState>> AllInit(KernelContext*, const KernelInitArg
   return ::arrow::internal::make_unique<BooleanAllImpl>();
 }
 
+// ----------------------------------------------------------------------
+// Index implementation
+
+struct IndexImpl : public ScalarAggregator {
+  explicit IndexImpl(IndexOptions options, int64_t seen, int64_t index)
+      : options(std::move(options)), seen{seen}, index{index} {}
+
+  Status Consume(KernelContext* ctx, const ExecBatch& batch) override {
+    // short-circuit
+    if (index >= 0 || !options.value->is_valid) {
+      return Status::OK();
+    }
+
+    const auto& data = *batch[0].array();
+    seen = data.length;
+    ARROW_ASSIGN_OR_RAISE(
+        auto result, CallFunction("equal", {data, options.value}, ctx->exec_context()));
```

Comment:
Looks not very efficient as "equal" scans the whole input array (and builds a bitmap) but we only want index of the first match.
I think we can leave it as follow up task to benchmark if visiting the array directly will give better performance.

Reply:
Yup, that is fair. I figured for now I didn't want to create a lot of generated code for this use case. It also at least helps that it'll short-circuit.